[CIRCLE-13803] Replace forked graphql client and improve error and response handling

zzak opened the pull request:
This PR is best viewed by commit, in particular 870de6c067928575320d937b2633bef8d8337723 and d637f12d5730f4344d6ea1b853da953d183e6bfb.

bradylill commented on client/client.go:
```diff
@@ -1,44 +1,135 @@
 
-	client := graphql.NewClient(endpoint)
+// RequestBody is used for serializing a request made up of a query and key/value pair of variables to send to the GraphQL server.
+type RequestBody struct {
```
What do you think about merging this struct with the `Request` struct below and adding the json serialization to that?

bradylill commented on client/client_test.go:
```diff
@@ -0,0 +1,208 @@
+		Data   map[string]interface{}
+		Errors []struct {
+			Message string
+		}
```
👍

dnephin commented on client/client.go:
```diff
@@ -1,44 +1,123 @@
+	defer func() {
+		err := res.Body.Close()
+		if err != nil {
+			log.Fatal(err)
```
This will `os.Exit()` if there is an error closing the body. That's probably not good. Closing the body is not really critical, especially in a CLI where the process is short lived. I would suggest either printing a warning  or ignoring it all together.

dnephin commented on client/client.go:
```diff
@@ -1,44 +1,123 @@
+	var buf bytes.Buffer
+	if _, err := io.Copy(&buf, res.Body); err != nil {
+		return errors.Wrap(err, "reading body")
+	}
```
Is the `io.Copy` necessary? I think we discussed this, I'm curious if you ran into any problems trying to remove it.

The `res.Body` is already a reader, so it can be passed to `NewDecoder` directly without buffering. If buffering is necessary for some reason https://godoc.org/bufio#NewReader would likely be a better option.

dnephin commented on cmd/namespace.go:
```diff
@@ -41,12 +41,12 @@ func createNamespace(cmd *cobra.Command, args []string) error {
 
 	// Only fall back to native graphql errors if there are no in-band errors.
-	if err != nil && (response == nil || len(response.Errors) == 0) {
+	if err != nil && (response == nil || len(response.Data.CreateNamespace.Errors) == 0) {
```
Aren't the errors in `response.Errors` ?

dnephin commented on api/api.go:
```diff
@@ -19,32 +19,51 @@ import (
+		messages = append(messages, errs[i].Message)
 	}
+
+	return strings.Join(messages, ": ")
```
Joining errors with a colon is a convention used to wrap an error with additional context as it passes through callers.

In this case it's being used to concat multiple errors, which seems very different. This could result in some very confusing errors. It might be better to join these errors with newlines or commas.

dnephin commented on api/api.go:
```diff
@@ -19,32 +19,51 @@ import (
+				Name string
+			}
+		} `json:"__schema"`
+	}
```
I believe this struct is missing an `Errors` field.

dnephin commented on api/api.go:
```diff
@@ -55,48 +74,198 @@ type ConfigResponse struct {
 	OutputYaml string
 
-	GQLResponseErrors
+	Errors GQLErrorsCollection
```
This `Errors` field is in the wrong spot. It needs to be at the top level of the `BuildConfigResponse` struct, as a sibling to `Data` field.

dnephin commented on api/api.go:
```diff
@@ -55,48 +74,198 @@ type ConfigResponse struct {
+		PublishOrb struct {
+			Orb Orb
+
+			Errors GQLErrorsCollection
```
Same with this `Errors` field, and all the ones below.

dnephin commented on api/api.go:
```diff
@@ -55,48 +74,198 @@ type ConfigResponse struct {
+			}
+		}
+	}
+}
```
missing errors

zzak commented:
@dnephin Thanks for your review! I've addressed most of these concerns in #150, so let's move the discussion there. In particular, some thoughts on in-band/out-of-band/runtime type errors being context specific.
